pegen/pegen.c

```diff
@@ -503,3 +503,64 @@ seq_flatten(Parser *p, asdl_seq *seqs)
 
     return flattened_seq;
 }
+
+expr_ty
+seq_to_dotted_name(Parser *p, asdl_seq *seq)
+{
+    /* Create a string of the form a.b.c like Python/ast.c alias_for_dotted_name does
+       for dotted names */
+    if (asdl_seq_LEN(seq) == 1) {
+        return asdl_seq_GET(seq, 0);
+    }
+
+    // Get length of dotted string
+    ssize_t len = 0;
+    for (int i = 0, l = asdl_seq_LEN(seq); i < l; i++) {
+        expr_ty current_expr = asdl_seq_GET(seq, i);
+        ssize_t current_len = PyUnicode_GET_LENGTH(current_expr->v.Name.id);
+        len += current_len + 1;
+    }
+    --len; // Last name does not have a dot
+
+    PyObject *str = PyBytes_FromStringAndSize(NULL, len);
+    if (!str) {
+        return NULL;
+    }
+
+    char *s = PyBytes_AS_STRING(str);
+    if (!s) {
+        return NULL;
+    }
+
+    for (int i = 0, l = asdl_seq_LEN(seq); i < l; i++) {
+        expr_ty current_expr = asdl_seq_GET(seq, i);
+        const char *current_name = PyUnicode_AsUTF8(current_expr->v.Name.id);
+        if (!current_name) {
+            return NULL;
+        }
+        strcpy(s, current_name);
+        s += strlen(current_name);
+        *s++ = '.';
```

Comment:
I'm not happy about how this "overshoots" the target and then backs up. I'd rather only add a dot *before* all items except the first.